core/common/src/main/java/alluxio/conf/path/PrefixPathMatcher.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.conf.path;
+
+import alluxio.AlluxioURI;
+
+import com.google.common.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Matched patterns are prefixes of a path, the path must be in the subtree of the prefixes.
+ *
+ * For example:
+ * /a/b/c matches /a/b, but
+ * /a/bc does not match /a/b because although /a/b is a prefix of /a/bc, /a/bc is not
+ * in the subtree of /a/b.
+ *
+ * Thread safety:
+ * Must be constructed in a single thread and published safely.
```

Comment:
I don't think it's possible to construct an object from multiple threads, or publish this object unsafely. This class looks perfectly threadsafe

Reply:
I mean it must be completely constructed in one thread before publishing. 

It might be published unsafely, for example, if the reference to this object is shared between two threads, one thread A is running the constructor, another thread B might already see the address of the allocated object although the constructor hasn't finished, so B might see a partially constructed object.

Reply:
According to https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5, that is not possible when all the fields are final

>final fields also allow programmers to implement thread-safe immutable objects without synchronization. A thread-safe immutable object is seen as immutable by all threads, even if a data race is used to pass references to the immutable object between threads. This can provide safety guarantees against misuse of an immutable class by incorrect or malicious code. final fields must be used correctly to provide a guarantee of immutability.

>An object is considered to be completely initialized when its constructor finishes. A thread that can only see a reference to an object after that object has been completely initialized is guaranteed to see the correctly initialized values for that object's final fields.

Reply:
nvm, all fields are final, JMM will prohibit reordering part of construction with the initial load of a reference to the object.

I removed the Thread safety note, it's completely thread safe.